lib/osut/utils.rb

```diff
@@ -1724,36 +1724,480 @@ def valid(surface = nil)
+    #   - sub[:head      ] # e.g. std 80" door + frame/buffers
+    #   - sub[:sill      ] # e.g. std 30" sill + frame/buffers
+    #   - sub[:height    ] # any sub surface height below "head"
+    #   - sub[:width     ] # e.g. 1.200 m
+    #   - sub[:offset    ] # centreline-to-centreline between subs (if array)
+    #   - sub[:centreline] # left/right of base surface centreline, e.g. -0.2 m
+    subs.each do |sub|
+      # Set-up unique sub parameters:
+      #   - Frame & Divider "width"
+      #   - minimum "clear glazing" limits
+      #   - buffers, etc.
+      id         = sub[:id]
+      frame      = 0
+      frame      = sub[:frame].frameWidth unless sub[:frame].nil?
+      frames     = 2 * frame
+      buffer     = frame + bfr
+      buffers    = 2 * buffer
+      dim        = 0.200 unless (3 * frame) > 0.200
+      dim        = 3 * frame if (3 * frame) > 0.200
+      glass      = dim - frames
+      min_sill   = buffer
+      min_head   = buffers + glass
+      max_head   = max_y - buffer
+      max_sill   = max_head - (buffers + glass)
+      min_ljamb  = buffer
+      max_ljamb  = max_x - (buffers + glass)
+      min_rjamb  = buffers + glass
+      max_rjamb  = max_y - buffer
+      max_height = max_y - buffers
+      max_width  = max_x - buffers
+
+      # Default sub surface "head" & "sill" height (unless user-specified).
+      typ_head = 2.032 # standard 80" door
+      typ_sill = 0.762 - buffer # standard 30" window sill
+
+      if sub.key?(:ratio)
+        typ_head = mid_y * (1 + sub[:ratio])     if sub[:ratio] > 0.75
+        typ_head = mid_y * (1 + sub[:ratio]) unless stype.downcase == "wall"
+        typ_sill = mid_y * (1 - sub[:ratio])     if sub[:ratio] > 0.75
+        typ_sill = mid_y * (1 - sub[:ratio]) unless stype.downcase == "wall"
+      end
+
+      # Log/reset "height" if beyond min/max.
+      if sub.key?(:height)
+        unless sub[:height].between?(glass, max_height)
+          sub[:height] = glass      if sub[:height] < glass
+          sub[:height] = max_height if sub[:height] > max_height
+          log(WRN, "Reset '#{id}' height to #{sub[:height]} m (#{mth})")
+        end
+      end
+
+      # Log/reset "head" height if beyond min/max.
+      if sub.key?(:head)
+        unless sub[:head].between?(min_head, max_head)
+          sub[:head] = typ_head
+          log(WRN, "Reset '#{id}' head height to #{sub[:head]} m (#{mth})")
+        end
+      end
+
+      # Log/reset "sill" height if beyond min/max.
+      if sub.key?(:sill)
+        unless sub[:sill].between?(min_sill, max_sill)
+          sub[:sill] = typ_sill
+          log(WRN, "Reset '#{id}' sill height to #{sub[:sill]} m (#{mth})")
+        end
+      end
+
+      # At this point, "head", "sill" and/or "height" have been tentatively
+      # validated (and/or have been corrected) independently from one another.
+      # Log/reset "head" & "sill" heights if conflicting.
+      if sub.key?(:head) && sub.key?(:sill) && sub[:head] < sub[:sill] + glass
```

Comment:
An example of one of many validation checks of user-set arguments. There are several options offered to users, e.g. sub surface (ww) _ratio_, _height_, _width_, _head_, _sill_, _spacing_ (if an array of sub surfaces), and hence many possible conflicts. Most of the code in _addSubs_ is a (longish), sequential series of validation checks. In many cases, the method resets conflicting input (while logging warnings). In other instances, it skips over an invalid combo of user arguments (while logging errors). This may be broken down into more than one method in the future ... maybe.